workshop/6-CustomerDataPlatforms/6.1-Segment.ipynb

```diff
@@ -0,0 +1,303 @@
+    "      \"style\": \"shirt\",\n",
+    "      \"description\": \"A classic look for the summer season.\",\n",
+    "      \"price\": 9.99,\n",
+    "      \"image\": \"http://recs.cloudfront.net/images/apparel/1.jpg\",\n",
+    "      \"featured\": \"true\"\n",
+    "    }\n",
+    "  },\n",
+    "  {\n",
+    "    \"product\": {\n",
+    "      \"id\": \"1\",\n",
+    "      \"url\": \"http://recs.cloudfront.net/#/product/1\",\n",
+    "      \"sk\": \"\",\n",
+    "      \"name\": \"Black Leather Backpack\",\n",
+    "      \"category\": \"accessories\",\n",
+    "      \"style\": \"bag\",\n",
+    "      \"description\": \"Our handmade leather backpack will look great at the office or out on the town.\",\n",
+    "      \"price\": 109.99,\n",
+    "      \"image\": \"http://recs.cloudfront.net/images/accessories/1.jpg\",\n",
+    "      \"featured\": \"true\"\n",
+    "    }\n",
+    "  },\n",
+    "\n",
+    "  ...\n",
+    "]\n",
+    "```\n",
+    "\n",
+    "Setting up a similar service in your deployment of Amazon Personalize will give you the ability to surface product recommendations directly into Segment to power other services that require real-time user personalization such as a mobile application or a web application.  \n",
+    "\n",
+    "The recommendation service deployed in the workshop queries a Personalize campaign with the `user_id` that you provide.  Personalize will return a list of recommended product IDs for the user.  The service then looks up the catalog ID for each product and decorates the response with the product URL, image, name, etc.  These attributes should be available in your catalog management system.  In the case of the Retail Demo Store, there is a catalog service that provides (simulated) catalog items for the workshop.  The advantage of a service like this is that your target services like Segment do not need to do an additional lookup in order to get the information they require in order to surface a recommendation to a user, such as product name, product category and the image associated with the product.\n",
+    "\n",
+    "Next, we will use the aforementioned service to pull our demo user recommendations and push them into Segment. We’re going to install Segment’s Python library, then create demo user profiles within Segment and pull in recommendations for each user from Amazon Personalize:\n"
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import sys\n",
+    "!{sys.executable} -m pip install analytics-python\n",
+    "import analytics\n",
+    "\n",
+    "analytics.write_key = segment_write_key\n",
+    "\n",
+    "for id in range(10, 20):\n",
+    "    analytics.identify('personalize_user_'+str(id), {\n",
+    "        'email': 'personalize_user_'+str(id)+'@example.com',\n",
+    "        'name': 'Example User'+str(id)\n",
+    "    })\n",
```

Comment:
Rather than using generic information for each user here, look up each user from the users service in the RDS so the name and email matches what they see in the Retail Demo Store. I suggest adding the persona from the user profile, if possible, too so that the workshop attendee can make more intuitive sense of what's being ingested into Personas. 

The user service IP address can be looked up using CloudMap/service discovery service using the following code.

```python
servicediscovery = boto3.client('servicediscovery')

response = servicediscovery.discover_instances(
    NamespaceName='retaildemostore.local',
    ServiceName='users',
    MaxResults=1,
    HealthStatus='HEALTHY'
)

users_service_instance = response['Instances'][0]['Attributes']['AWS_INSTANCE_IPV4']
print('Users Service Instance IP: {}'.format(users_service_instance))
```
Then you can look up the profile for each user in your `for` loop like this:

```python
response = requests.get('http://{}/users/id/{}'.format(users_service_instance, id))
user = response.json()
name = f'{user["first_name"]} {user["last_name"]}'
email = user['email']
persona = user['persona']
```

If the concern is possibly polluting a workshop attendee's production workspace, you could prefix or suffix each name with something like "(Personalize workshop)" or something.

Here is example of a response from the users service.

```javascript
{
    "addresses": [
        {
            "address1": "2058 Alison Locks Apt. 412",
            "address2": "",
            "city": "Shellyview",
            "country": "US",
            "default": true,
            "first_name": "Tiffany",
            "last_name": "Mcclure",
            "state": "TN",
            "zipcode": "37097"
        }
    ],
    "age": 41,
    "email": "tiffany.mcclure@example.com",
    "first_name": "Tiffany",
    "gender": "F",
    "id": "256",
    "last_name": "Mcclure",
    "persona": "accessories_jewelry_apparel",
    "username": "user256"
}
```

Reply:
Done.  Great suggestion.